### ISSUE TITLE
Follow links when disconnecting

### ISSUE THREAD

**Issue by ghost** [dotnet/aspnetcore#9964]
As soon as you have lost contact with blazor.server.js, any transition to the links of the prejudice causes the error 'blazor.server.js:1 Uncaught (in promise) Error: Cannot send data if the connection is not in the 'Connected' State.'

I suggest in the absence of communication with the server to perform the usual transition by reference, without using server rendering. In this case, you do not need to display a window about disconnect, and the normal reconnection will be performed with the page reloading.

**Comment by SteveSandersonMS** [comment]
@pranavkm This is a good idea. Since you're working on the navigation handling, do you know how feasible it would be to add the further rule "if the client is unable to send the navigation event to the server for the router to handle, just do a full page load directly"?

**Comment by SteveSandersonMS** [comment]
BTW I don't mind whether it still logs an error too. Perhaps it's best if it does, since it is an error case. Real-world users would not see anything go into the console anyway.

**Comment by cores-system** [comment]
@SteveSandersonMS , The ability to catch the error is needed so that the user knows why the button / form/ etc.
But if a user is a guest viewing a blog / news for example https://devblogs.microsoft.com/aspnet/, then it is more logical to make a transition to a new page.

That is, the navigation itself should not depend on whether the user is connected to a socket or not.

**Comment by SteveSandersonMS** [comment]
Navigation has to depend on the socket connection, because the .NET code in the Router makes the choice about how navigation is handled. My suggestion above only applies in the case where the connection is broken and we are just trying to recover the situation for the user.

**Comment by cores-system** [comment]
I mean links without logic `<a href="/conter">conter</a>`, during disconnect, we can make a simple transition.

**Comment by SteveSandersonMS** [comment]
Yes, we're saying the same thing :)

**Comment by SteveSandersonMS** [comment]
Pranav, this is related to the routing stuff you're currently working on (and disconnection you previously worked on). Do you think [this](https://github.com/aspnet/AspNetCore/issues/9964#issuecomment-489549079) is reasonable to do?

**Comment by pranavkm** [comment]
Just saw @SteveSandersonMS's comments further up, but this looks like a reasonable implementation. SignalR can tell us of the connection's state and that can be used to determine what to do here.

**Comment by barahonajmc** [comment]
This sound interesting, being able to click on links even if we are disconnected and specially those links that does not even depend of Blazor because they redirect you to outside places but I have some thoughts:


Does this only apply when the user clicks on a link just before you lose the connection? because:

- Currently a disconnection blocks the UI completely and I believe that is on purpose as the server could had being save the user state and if the user continues modifying it then after reconnection there will be some kind of rollback in the data

**Comment by RushilJalal** [comment]
I will be glad to help!

**Comment by oguzozshn** [comment]
I've opened #68562 as a draft with a possible fix for this. It's a draft rather than a normal PR because there's a design question here that I'd rather settle before the code goes any further.

**Why it fails today**

`NavigationManager.attachToEventDelegator` intercepts internal link clicks as long as an interactive router is registered, without considering whether that router can be reached. `performInternalNavigation` then ends in `notifyLocationChanged`, which dispatches to .NET over the circuit — that's where `Cannot send data if the connection is not in the 'Connected' State` comes from. Since `handleClickForNavigationInterception` has already called `preventDefault()`, the browser's own navigation is suppressed as well, so the user stays on the page with only a console error.

**The approach**

The mechanism needed here already exists: `currentPageLoadMechanism()` can return `'serverside-fullpageload'`, and `navigateToCore` routes that to `performExternalNavigation`. So the change mostly widens the condition that selects it — the click handler returns early without `preventDefault()`, letting the browser do a normal page load, which follows the link and gives the app a chance to establish a new circuit. WebAssembly never registers a connection checker and so is unaffected. 38 lines of product code, no public API change.

**The design question**

A full page load discards circuit state, which is what the reconnection UI exists to avoid. My reasoning for accepting that here is that this path is only reached after the user clicked a link to another page, so the current page's state is being left behind either way — and today the alternative isn't "state is preserved", it's a swallowed navigation plus an unhandled error.

If you'd prefer this to be opt-in, or gated on the reconnection attempts having been exhausted rather than on the connection simply not being `Connected` at that moment, I'm happy to rework it — that's the main thing I'd like guidance on.

I've deliberately left back/forward (`onPopState`) out of scope; it reaches the same interop call while disconnected, but the desired behavior there seems less clear-cut.
